pkg/pgclient/config.go

```diff
@@ -135,70 +142,65 @@ func (cfg *Config) GetConnectionStr() string {
 	return cfg.DbUri
 }
 
-func (cfg *Config) GetNumConnections() (min int, max int, numCopiers int, err error) {
-	maxProcs := runtime.GOMAXPROCS(-1)
-	if cfg.WriteConnections < 0 {
-		return 0, 0, 0, fmt.Errorf("invalid number of writer connections %v, cannot be a negative number", cfg.WriteConnections)
-	}
-
-	conn, err := pgx.Connect(context.Background(), cfg.GetConnectionStr())
+// GetPoolSize returns the max pool size based on the max_connections allowed by database and the defaultFraction.
+// Arg inputPoolSize is the pool size provided in the CLI flag.
+func (cfg *Config) GetPoolSize(poolName string, defaultFraction float64, inputPoolSize int) (int, error) {
+	maxConns, err := cfg.maxConn()
 	if err != nil {
-		return 0, 0, 0, err
+		return 0, fmt.Errorf("max connections: %w", err)
 	}
-	defer func() { _ = conn.Close(context.Background()) }()
-
-	max = cfg.MaxConnections
-	if max < 1 {
-		var maxStr string
-		row := conn.QueryRow(context.Background(), "SHOW max_connections")
-		err = row.Scan(&maxStr)
-		if err != nil {
-			return 0, 0, 0, err
-		}
-		max, err = strconv.Atoi(maxStr)
-		if err != nil {
-			log.Warn("err", err, "msg", "invalid value from postgres max_connections")
-			max = 100
-		}
 
-		//In HA setups
+	switch {
+	case inputPoolSize == defaultPoolSize:
+		// For the default case, we need to take up defaultFraction of allowed connections.
+		poolSize := float64(maxConns) * defaultFraction
 		if cfg.UsesHA {
-			max = max / 2
+			poolSize /= 2
 		}
+		return int(poolSize), nil
+	case inputPoolSize < MinPoolSize:
+		return 0, fmt.Errorf("%s pool size canot be less than %d: received %d", poolName, MinPoolSize, inputPoolSize)
+	case inputPoolSize > maxConns:
+		return 0, fmt.Errorf("%s pool size canot be greater than the 'max_connections' allowed by the database", poolName)
+	default:
+	}
+	if cfg.UsesHA {
```

Comment:
Nit: doing this check two times and you can do it just once. Improve if you can but lets not waste time on it.